### ISSUE TITLE
Receiving email without being logged in

### ISSUE THREAD

**Issue by lupine** [emersion/go-smtp#8]
Hi there,

Thanks for your work on go-imap and go-smtp! I'm attempting to use them for an email-related project, and while it seems I can use go-smtp to implement a message submission port well enough, I *can't* use it to accept email from the wider Internet (e.g., the kind of thing Exim usually does when configured as an internet site).

An SMTP transaction of this kind would generally look like:

```
EHLO example.com
MAIL FROM:...
RCPT TO:...
DATA
...
.
QUIT
```

Would you be willing to consider a PR that allows this behaviour to be optionally enabled? Something like:

```go
be := mybackend.NewSubmission()
unauthenticatedBE := mybackend.NewSMTP()

server := smtp.NewServer(backend)
server.UnauthenticatedBackend = unauthenticatedBE
```

If the `UnauthenticatedBackend` is nil, we can continue to reject as we do presently. If it's set, login is optional. If a message comes through without a login happening, the `UnauthenticatedBackend` handles it. 

We could also add:

```go
server.LoginDisabled = true
```

This would *disable* the AUTH command completely, which would make things a bit cleaner from my side :)

Happy to do the work for this, I just wanted to open a line of communication about it before starting! Let me know.

**Comment by emersion** [comment]
Hi! That's a good idea indeed. Maybe we could just add a `AnonymousLogin() (User, error)` function to the `Backend` interface?

**Comment by lupine** [comment]
Hmm, so we'd call it if we get to `MAIL` without an `AUTH PLAIN` being received?

The change wouldn't be backward-compatible - is that a concern?

**Comment by emersion** [comment]
>Hmm, so we'd call it if we get to MAIL without an AUTH PLAIN being received?

That's the idea. We could also add `Send` to `Backend` directly, but that would duplicate the function and be confusing.

>The change wouldn't be backward-compatible - is that a concern?

No, the README explicitly states that this package is unstable. I'll consider making it stable and publishing a v1 after this change.

**Comment by lupine** [comment]
Great, I'll see if I can find the time to push a branch in the next day or two for review :).

Do you agree on the `server.LoginDisabled` attribute? If I know ahead of time that I won't be accepting logins, it seems to make sense to not advertise the capability at all?

**Comment by emersion** [comment]
Hmm, I wonder if it makes more sense for this to be a property of the backend. We could make `Login` optional (by moving it to another interface) for instance. Still not sure about it, just throwing ideas around.

But yes, we definitely want that.

**Comment by lupine** [comment]
I've implemented it as a property on the server in https://github.com/emersion/go-smtp/pull/9 but I'm open to ideas. Let's discuss it more over there!

**Comment by emersion** [comment]
Fixed in #9 :tada:
